### DISCUSSION TITLE
switch-to-buffer

### DISCUSSION THREAD

**Discussion by snan** [minad/consult#892] [Q&A] [answered]
I'm using vertico, consult, consult embark, and orderless and it's OK but for `switch-to-buffer` specifically it kind of bugs me that the order isn't connected to the last-used-order (the order that `list-all-buffers` uses). I'm like "I just used that buffer three seconds ago and now I need to search for it?!"

Help please? 🙏🏻♥

**Comment by minad** [comment]
On 11/26/23 16:04, snan wrote:
> I'm using vertico, consult, consult embark, and orderless and it's OK
> but for |switch-to-buffer| specifically it kind of bugs me that the
> order isn't connected to the last-used-order (the order that
> |list-all-buffers| uses). I'm like "I just used that buffer three
> seconds ago and now I need to search for it?!"

Hi, do you mean `consult-buffer` or `switch-to-buffer`? For me both
`consult-buffer` and `switch-to-buffer` work well with a reasonable
order and so far I've not gotten reports about problems with the
candidate order. Could you please provide a minimal configuration based
on emacs -Q, see https://github.com/minad/consult#bug-reports? Could you
be more precise about your observations and provide a precise recipe
such that I can reproduce this? Thanks!

**Comment by snan** [comment]
Hi!

What I mean specifically is that when I'm using `switch-to-buffer` it's great if I'm looking for the very most recently used buffer, but if I'm looking for something further back it's utterly lost and I either need to search for it by typing letters from it, or give up with `C-g` and instead use `list-buffers`, which does list the buffers in the order they were used so I can find my recently-but-not-the-very-recentliest buffer.

I'm 💯 OK with the `switch-to-buffer` jumping around a bit when I start typing as it starts listing the buffers in the most likely order based on what I've typed, I just want the very first before-I've-typed-anything order to be recency based so that I can arrow-key down to what I want. Also it's because I forget the name of the buffer and being able to see it would prevent that.

Here is my config ♥

    (straight-use-package 'use-package)
    (use-package vertico
      :straight (:files (:defaults "extensions/*"))
      :init
      (vertico-mode))
    (straight-use-package 'savehist)
    (savehist-mode)
    (straight-use-package 'consult)
    (straight-use-package 'embark-consult)

    (defun crm-indicator (args)
      (cons (format "[CRM%s] %s"
                    (replace-regexp-in-string
                     "\\`\\[.*?]\\*\\|\\[.*?]\\*\\'" ""
                     crm-separator)
                    (car args))
            (cdr args)))
    (advice-add #'completing-read-multiple :filter-args #'crm-indicator)

    (setq minibuffer-prompt-properties
          '(read-only t cursor-intangible t face minibuffer-prompt))
    (add-hook 'minibuffer-setup-hook #'cursor-intangible-mode)

    (straight-use-package 'orderless)

    (setq completion-styles '(orderless basic)
          orderless-matching-styles '(orderless-literal orderless-regexp orderless-prefixes orderless-initialism orderless-flex)
          completion-category-defaults nil
          completion-category-overrides '((file (styles partial-completion))))

    (use-package marginalia
      :straight t
      :bind (:map minibuffer-local-map
             ("M-A" . marginalia-cycle))
      :init
      (marginalia-mode))

    (use-package vertico-directory
      :after vertico
      :ensure nil
      :bind (:map vertico-map
                  ("RET" . vertico-directory-enter)
                  ("DEL" . vertico-directory-delete-char)
                  ("M-DEL" . vertico-directory-delete-word))
      :hook (rfn-eshadow-update-overlay . vertico-directory-tidy))

    (use-package embark
      :straight t
      :ensure t
      :bind
      :init
      (setq prefix-help-command #'embark-prefix-help-command
    	embark-prompter #'embark-completing-read-prompter))

    (setq completion-in-region-function
          (lambda (&rest args)
            (apply (if vertico-mode
                       #'consult-completion-in-region
                     #'completion--in-region)
                   args)))

**Comment by minad** [comment]
> What I mean specifically is that when I'm using |switch-to-buffer| it's
> great if I'm looking for the very most recently used buffer, but if I'm
> looking for something further back it's utterly lost and I either need
> to search for it by typing letters from it, or give up with |C-g| and
> instead use |list-buffers|, which does list the buffers in the order
> they were used so I can find my recently-but-not-the-very-recentliest
> buffer.
> 
> I'm 💯 OK with the |switch-to-buffer| jumping around a bit when I start
> typing as it starts listing the buffers in the most likely order based
> on what I've typed, I just want the very first
> before-I've-typed-anything order to be recency based so that I can
> arrow-key down to what I want. Also it's because I forget the name of
> the buffer and being able to see it would prevent that.

I don't observe the described problem. In both `consult-buffer` and
`switch-to-buffer` the buffers are shown in the order returned by
`buffer-list`, sorted by recency.

**Comment by snan** [comment]
I still have the issue 🤷🏻‍♀️
But `consult-buffer` which I had never heard of doesn't have the problem so I'll bind to it, thank you 🙏🏻

**Reply by snan** [reply]
It sorta previews the buffer but that might be something I could learn to appreciate and even enjoy. More importantly, it shows the buffers in the right order. 🙏🏻

**Comment by daanturo** [comment]
I also have this issue.

The last buffer not being on the top is always reproducible for [ein](https://github.com/millejoh/emacs-ipython-notebook)'s running notebook buffers.

Steps to reproduce from `emacs -Q`:

1. Add MELPA, install `consult`, `ein`.
2. `M-x fido-vertical-mode` (for being quickly, else `vertico`)
3. `M-x ein:run`
4. Open an exising jupyter notebook or `[New Notebook]`
5. Switch to any other buffer
6. Run `consult-buffer`

Expected: `*ein: http://127.0.0.1:8889/Untitled.ipynb*` on top of the buffer list.
Like the built-in `switch-to-buffer`:
![spectacle-my-screenshot-window_20240202_225816](https://github.com/minad/consult/assets/32123754/9f43b983-5787-4a37-bc64-c4e0903a60e1)


Observed: for `consult-buffer` it's buried quite far below:
![spectacle-my-screenshot-window_20240202_230025](https://github.com/minad/consult/assets/32123754/bfdac324-d725-4ed7-86d1-830eef6ead20)

By making `other-buffer` the head of the (sorted) buffer list, I can achieve the result after switching buffer once: https://github.com/daanturo/consult/commit/04aef58bcda605dbb0b1a5a943c8fa735593a075, but `switch-to-buffer` also does something that's also applicable to the second switch (to a 3rd buffer).

**Reply by minad** [reply]
Can you please create a minimal reproducible example only with Consult and Vertico starting from emacs -Q? If it is a problem with the buffer switcher, it should be reproducible without EIN or other packages.

**Comment by rudolf-adamkovic** [comment, marked as answer]
After suffering from this problem for years, and almost switching to `icomplete`, I think I found something that works.  I studied `switch-to-buffer` (and the functions it calls, all the way down to C) and then I studied `consult-buffer`.  The code blow fixes the problem for me, resulting in completions that are usable, as in "matching `list-buffers`", and not only in a minimal reproducer but throughout a heavy workday with 50 to 100 buffers open.  (Tested for one entire day.)

```
(with-eval-after-load 'vertico-multiform
  (add-to-list 'vertico-multiform-categories
	       '(buffer (vertico-sort-function . nil))))
```

Of course, `(add-hook 'vertico-mode-hook #'vertico-multiform-mode)` is a prerequisite.
